### PR TITLE
Improve performance of windows build script on rebuild

### PR DESCRIPTION
**Pull request by spacehamster**
Amended windows build script to skip redundant work on rebuild

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -162,11 +162,11 @@ sudo make install
 1. Build Memento:
     ```
     cd Memento
-    ./windows/build.sh x86_64
+    ./windows/build.sh x86_64 release
     ```
 1. The resulting file will be in
     ```
-    build/Memento_x86_64
+    build/release_x86_64\Memento
```
Comment (ripose-jp) [inline README.md:169]: Change to forward slash

File: windows/build.sh [hunk 2]
```diff
@@ -25,31 +34,40 @@ PREFIX=/mingw32
...
+root=$(pwd)
+
 # build Memento
-mkdir -p build
-cd build
-$PREFIX/bin/cmake -DCMAKE_BUILD_TYPE=Release ${CMAKE_ARGS} ..
+mkdir -p build/${build_type}_${arch}
+cd build/${build_type}_${arch}
+if [ ! -f ./build.ninja ]; then
+$PREFIX/bin/cmake -G "Ninja" -DCMAKE_BUILD_TYPE=$build_type ${CMAKE_ARGS} $root
```
Comment (ripose-jp) [inline windows/build.sh:43]: Add quotations around `$root`. It may have spaces in the path.

File: windows/build.sh [hunk 3]
```diff
@@ -25,31 +34,40 @@ PREFIX=/mingw32
...
-cp src/memento.exe Memento_$arch
-rm -rf Memento_$arch/memento_debug.exe
-cp src/memento_debug.exe Memento_$arch
-cp -r ../dic Memento_$arch
+mkdir -p Memento
+rm -rf Memento/memento.exe
+cp src/memento.exe Memento
+rm -rf Memento/memento_debug.exe
+cp src/memento_debug.exe Memento
+cp -r $root/dic Memento
```
Comment (ripose-jp) [inline windows/build.sh:53]: Add `$root` quotations

File: windows/build.sh [hunk 4]
```diff
@@ -25,31 +34,40 @@ PREFIX=/mingw32
...
-$PREFIX/bin/cmake -DCMAKE_BUILD_TYPE=Release ${CMAKE_ARGS} ..
+mkdir -p build/${build_type}_${arch}
+cd build/${build_type}_${arch}
+if [ ! -f ./build.ninja ]; then
+$PREFIX/bin/cmake -G "Ninja" -DCMAKE_BUILD_TYPE=$build_type ${CMAKE_ARGS} $root
+fi
 $PREFIX/bin/cmake --build . -- -j$(nproc)
 
 # move DLLs and exe to a new directory
-mkdir -p Memento_$arch
```
Comment (ripose-jp) [inline windows/build.sh:35]: Please keep the `$arch` stuff

File: windows/build.sh [hunk 5]
```diff
@@ -25,31 +34,40 @@ PREFIX=/mingw32
...
+    cp _deps/libmocr-src/build/libmocr.dll Memento
+    cp _deps/libmocr-src/build/libmocr++.dll Memento
 fi
 
-python3 ../windows/mingw-bundledlls.py --copy ./Memento_$arch/memento.exe
-cp $PREFIX/bin/libssl-*.dll ./Memento_$arch
+cd Memento
+
+if [ ! -f ./Qt6Core.dll ]; then
+    python3 $root/windows/mingw-bundledlls.py --copy ./memento.exe
```
Comment (ripose-jp) [inline windows/build.sh:63]: `$root` quotations

**Review by ripose-jp — CHANGES_REQUESTED** [review]

**Comment by ripose-jp** [comment]
[Please update your commit message in accordance with the guidelines.](https://github.com/ripose-jp/Memento/blob/master/CONTRIBUTING.md#commits)

Not that I'm necessarily against this change, but currently it's going to break other scripts. You'll also need to modify `build-installer.sh` as well since it depends on `build.sh`.

I think you might be making this change since you have a suboptimal development environment. If you're using VSCode what I like to do is the following:

* Add Msys2-MINGW64's bin folder to my PATH: `C:\msys64\mingw64\bin`
* [Add Msys2-MINGW64 as a terminal in VSCode](https://stackoverflow.com/questions/45836650/how-do-i-integrate-msys2-shell-into-visual-studio-code-on-window#answer-68253833)
* [Set the default terminal profile to Msys2](https://stackoverflow.com/questions/44435697/change-the-default-terminal-in-visual-studio-code)
* Install the [Microsoft C/C++ Extensions](https://marketplace.visualstudio.com/items?itemName=ms-vscode.cpptools-extension-pack) pack, this will include the CMake plugin
* Install [Native Debug](https://marketplace.visualstudio.com/items?itemName=webfreak.debug) plugin

After doing this, you shouldn't need to run this build script anymore unless you want to make portable binaries. You should be able to just click the **Build** button in VSCode. gdb and everything else should work as well once you set up a `launch.json` using Native Debug.

**Comment by spacehamster** [comment]
> I think you might be making this change since you have a suboptimal development environment.

Yes, I had been using VSCode without using the CMake Plugin. This solves the main issue I was having. It still rebuilds from when switching build_type (debug/release), but that doesn't happen often to bother me. Do you want me to make the suggested changes or just close the pull request?

Also looking briefly, both Native Debug and Microsoft C/C++ Extensions support debugging with Msys2, is there an advantage to one to the other? I'd just been using Microsoft C/C++ Extensions.

**Comment by ripose-jp** [comment]
>It still rebuilds from when switching build type (debug/release)

It has to. The binary generated by the compiler when given `-O0` and `-O3` are completely different.

>Do you want me to make the suggested changes or just close the pull request?

I really just use these build scripts to create portable binaries. Adding features useful for development doesn't seem too useful in my opinion, so I lean towards yes.

>Microsoft C/C++ Extensions support debugging with Msys2

That's what they want you to think, but I have never ever gotten visual debug on VSCode to work on Windows except with Native Debug. It's the only reason I recommend it, since I know at this point the problem with Microsoft's extension isn't just user error.

**Pull request closed without merging**